lib/bundler/templates/Executable.bundler

```diff
@@ -38,10 +38,39 @@ m = Module.new do
   end
 
   def gemfile
-    gemfile = ENV["BUNDLE_GEMFILE"]
-    return gemfile if gemfile && !gemfile.empty?
+    gemfile = find_gemfile
+    return unless gemfile
+    File.expand_path(gemfile)
+  end
+
+  def find_gemfile
+    given = ENV["BUNDLE_GEMFILE"]
+    return given if given && !given.empty?
+    find_file(*gemfile_names)
+  end
 
-    File.expand_path("../<%= relative_gemfile_path %>", __FILE__)
+  def gemfile_names
+    ["Gemfile", "gems.rb"]
+  end
+
+  def find_file(*names)
+    search_up(*names) do |filename|
+      return filename if File.file?(filename)
+    end
+  end
+
+  def search_up(*names)
+    previous = nil
+    current  = File.expand_path(Dir.pwd)
```

Comment:
should this use the binstub directory instead of the current working directory?

Reply:
It should be the current working directory. If a Gemfile is not found in that current directory it should continue searching upwards to find a Gemfile, similar to [`SharedHelpers.find_gemfile`](https://github.com/bundler/bundler/blob/cecdfdb5b2a76133b0a83093ff6d80d1ffd97b46/lib/bundler/shared_helpers.rb#L240-L246).

The problem that this PR addresses is that when a bundler binstub is created and then used to install other gem executables, the bundler binstub currently stores the path of the Gemfile used in the last successful `bundle install`. Trying to run `bundle exec` when bundling against multiple gemfiles could result in a `Gem::Exception` or `LoadError` because bundler may not be using the correct Gemfile path.

From #6162:

> ...`bundle exec foo` should _always_ find the Gemfile of the current directory (when BUNDLE_GEMFILE is not set) and then run the `foo` for that Gemfile.

Reply:
but this isn't `bundle exec foo`, this is `bin/foo` ?

Reply:
I'm not quite sure I follow... 

This particular template, `lib/bundler/templates/Executable.bundler`, sets up a special bundler binstub that ensures the correct version of bundler is activated ([commit](https://github.com/bundler/bundler/commit/01fe509b27b317dd4d640beb9140ab31331ea349)). When `bundle exec foo` is run, this binstub is executed and [loads a version of bundler](https://github.com/bundler/bundler/blob/ba49ed283fa20d313f95cdefcd32f8f82a786c9f/lib/bundler/templates/Executable.bundler#L104).

The bundler binstub sets `ENV['BUNDLE_GEMFILE']`, which is then passed along to bundler. At this point, if `ENV['BUNDLE_GEMFILE']` is pointing to a gemfile that doesn't include `foo`, a `Gem::Exception` or `LoadError` may be raised even if `bin/foo` exists.

Not sure if this helps, but it may be easier to follow through this [Bundler repro script for bundler issue 6154](https://gist.github.com/eanlain/83f97f919701c864fb68dfdcf813cf5a).

Reply:
correct, but you quoted me documentation for `bundle exec`, and how _it_ should find a gemfile. Shouldn't the `bundle binstubs` binstubs _always_ refer to the Gemfile they were generated with, even if executed from another directory?

Reply:
I see what you're saying... yes, I agree, except in the case of bundler itself (`bin/bundle`). The binstub for bundle shouldn't refer to the Gemfile that it was generated with as it runs into the previously mentioned issues.

Reply:
the only thing is that the other binstubs all use the same logic as the `bundle` binstubs when it comes to this, iirc